error.go

```diff
@@ -54,6 +55,113 @@ func RootCause(err error) error {
 	}
 }
 
+// errf is a version of fmt.Errorf with support for a chain of multiple
+// formatted error messages.
+//
+// After msg, N arguments are consumed as formatting arguments for that
+// message, where N is the number of % symbols in msg. Following that, another
+// string may be added to become the next error in the chain. Each new error
+// is the `cause()` for the prior error.
+//
+//   err := errf(
+//     "could not process %v", thing,
+//     "name %q is invalid", thing.Name,
+//  )
+//  fmt.Println(err)  // could not process Thing: name Foo is invalid
+//  fmt.Println(RootCause(err))  // name Foo is invalid
+//
+// In place of a string, the last error can be another error, in which case it
+// will be treated as the cause of the prior error chain.
+//
+//   errf(
+//     "could not process %v", thing,
+//     "date %q could not be parsed", thing.Date,
+//     parseError,
+//  )
+func errf(msg string, args ...interface{}) error {
+	// To ensure that errf controls all calls to buildErrf, we make it a
+	// closure instead of its own function. This allows buildErrf to assume
+	// that len(args) > 0 for all calls.
+	var buildErrf func([]interface{}) error
```

Comment:
I can appreciate the utility of recursion for building a nested structure but I also feel like it would be possible to make this pretty readable if you just used iteration instead of a closure + recursion.

Reply:
I specifically wrote an iterative version of this because I felt that
the extra function needed for the recursive approach was a bit icky.
I went back to this because the iterative approach becomes difficult to
follow in terms of exactly how the information flows, since you have to
do it in two stages:

- `[]interface{}` to `[]string` based on formatting arguments, and a
  `rootCause error`.
- Build up the chain in reverse from `[]string` with `rootCause` as the
  cause of the final error in the chain.

I can switch to it if you'd like, but I think for someone reading this
code, the flow of information there is less clear.